src/gui/IconDownloaderDialog.cpp

```diff
@@ -0,0 +1,186 @@
+    QEventLoop loop;
+    connect(downloader.data(), SIGNAL(iconReceived(const QImage&, Entry*)), &loop, SLOT(quit()));
+    loop.exec();
+#endif
+}
+
+bool IconDownloaderDialog::downloadAllFavicons(const QSharedPointer<Database>& database, const QList<Entry*>& entries)
+{
+    m_db = database;
+
+    // Set progress bar
+    int maximum = 0;
+    for (const auto& e : entries) {
+        if (!e->url().isEmpty() && e->iconUuid().isNull()) {
+            ++maximum;
+        }
+    }
+    m_ui->progressBar->setMinimum(0);
+    m_ui->progressBar->setMaximum(maximum);
+
+    for (const auto& e : entries) {
+        if (!e->url().isEmpty() && e->iconUuid().isNull()) {
+            QFuture<void> fut = QtConcurrent::run(this, &IconDownloaderDialog::downloadFavicon, database, e);
+            m_futureList.addFuture(fut);
+        }
+    }
+
+    m_futureList.waitForFinished();
+    m_ui->abortButton->setEnabled(false);
+    return true;
+}
+
+void IconDownloaderDialog::iconReceived(const QImage& icon, Entry* entry)
+{
+    QMutexLocker locker(&m_mutex);
+    m_ui->progressBar->setValue(m_ui->progressBar->value() + 1);
+    m_ui->label->setText(tr("Downloading favicon %1 of %n…", 0, m_ui->progressBar->maximum()).arg(m_ui->progressBar->value()));
+
+    // Update the table
+    for (int i = 0; i < m_dataModel->rowCount(); ++i) {
+        if (m_dataModel->item(i, 0)->text() == entry->url()) {
+            m_dataModel->item(i, 1)->setText(icon.isNull() ? tr("Error") : tr("Done"));
+        }
+    }
+
+    if (!addCustomIcon(icon, entry)) {
+        emit messageEditEntry(tr("Custom icon already exists"), MessageWidget::Information);
+    }
+}
+
+bool IconDownloaderDialog::addCustomIcon(const QImage& icon, Entry* entry)
+{
+    bool added = false;
+    if (m_db && !icon.isNull()) {
+        // Don't add an icon larger than 128x128, but retain original size if smaller
+        auto scaledicon = icon;
+        if (icon.width() > 128 || icon.height() > 128) {
+            scaledicon = icon.scaled(128, 128);
+        }
+
+        QUuid uuid = m_db->metadata()->findCustomIcon(scaledicon);
+        if (uuid.isNull()) {
+            uuid = QUuid::createUuid();
+            m_db->metadata()->addCustomIcon(uuid, scaledicon);
+            m_customIconModel->setIcons(m_db->metadata()->customIconsScaledPixmaps(),
+                                        m_db->metadata()->customIconsOrder());
+            added = true;
+        }
+
+        if (entry) {
+            entry->setIcon(uuid);
+        }
+
+        emit entryUpdated();
+    }
+
+    return added;
+}
+
+void IconDownloaderDialog::fallbackNotEnabled()
+{
+     emit messageEditEntry(
+        tr("Unable to fetch favicon.") + "\n"
+            + tr("You can enable the DuckDuckGo website icon service under Tools -> Settings -> Security"),
+        MessageWidget::Error);
+}
+
+void IconDownloaderDialog::iconError()
+{
+     emit messageEditEntry(tr("Unable to fetch favicon."), MessageWidget::Error);
```

Comment:
This would look better at the "Status" column. When there are a lot of icons being download at once, I have no idea which one this message is referring to. [2]